fake_joint_driver/src/fake_joint_driver.cpp

```diff
@@ -25,103 +24,131 @@ FakeJointDriver::FakeJointDriver(void)
+  cmd_vel_.resize(joint_names_.size(), 0.0);
+  act_dis_.resize(joint_names_.size());
+  act_vel_.resize(joint_names_.size());
+  act_eff_.resize(joint_names_.size());
 
   // Set start position
-  for (auto it=start_position_map.begin(); it!=start_position_map.end(); it++)
-  {
-    for (auto i=0; i<joint_names_.size(); i++)
-    {
-      if (joint_names_[i] == it->first)
-      {
-        act_dis[i] = it->second;
-        cmd_dis[i] = it->second;
+  for (auto it = start_position_map.begin(); it != start_position_map.end();
+       ++it) {
+    for (auto i = 0; i < joint_names_.size(); ++i) {
+      if (joint_names_[i] == it->first) {
+        act_dis_[i] = it->second;
+        cmd_dis_[i] = it->second;
+        cmd_dis_old_[i] = cmd_dis_[i];
       }
     }
   }
 
   // Create joint_state_interface and position_joint_interface
-  for (int i = 0; i< joint_names_.size(); i++)
-  {
+  for (int i = 0; i < joint_names_.size(); ++i) {
     ROS_DEBUG_STREAM("joint[" << i << "]:" << joint_names_[i]);
     // Connect and register the joint_state_interface
-    hardware_interface::JointStateHandle state_handle(joint_names_[i], &act_dis[i], &act_vel[i], &act_eff[i]);
-    joint_state_interface.registerHandle(state_handle);
+    hardware_interface::JointStateHandle state_handle(
+        joint_names_[i], &act_dis_[i], &act_vel_[i], &act_eff_[i]);
+    joint_state_interface_.registerHandle(state_handle);
 
     // Connect and register the position_joint_interface
-    hardware_interface::JointHandle pos_handle(joint_state_interface.getHandle(joint_names_[i]), &cmd_dis[i]);
-    position_joint_interface.registerHandle(pos_handle);
+    hardware_interface::JointHandle pos_handle(
+        joint_state_interface_.getHandle(joint_names_[i]), &cmd_dis_[i]);
+    position_joint_interface_.registerHandle(pos_handle);
+
+    // Connect and register the velocity_joint_interface
+    hardware_interface::JointHandle vel_handle(
+        joint_state_interface_.getHandle(joint_names_[i]), &cmd_vel_[i]);
+    velocity_joint_interface_.registerHandle(vel_handle);
   }
-  registerInterface(&joint_state_interface);
-  registerInterface(&position_joint_interface);
+  registerInterface(&joint_state_interface_);
+  registerInterface(&position_joint_interface_);
+  registerInterface(&velocity_joint_interface_);
 }
 
-FakeJointDriver::~FakeJointDriver()
-{
+/**
+ * @brief Update function to call all of the update function of motors
+ */
+void FakeJointDriver::update(ros::Duration period) {
+
+  // check if non zero velocity command is sent
+  double vel_sum = 0.0;
+  std::for_each(cmd_vel_.begin(), cmd_vel_.end(),
+                [&vel_sum](double a) { return vel_sum += std::abs(a); });
+
+  // if velocity commands are zero - use position interface
+  if (vel_sum == 0.0) {
+    // TODO enforce position and velocity limits
+    // only do loopback
+    act_dis_ = cmd_dis_;
+    std::fill(cmd_vel_.begin(), cmd_vel_.end(), 0.0);
+    act_vel_ = cmd_vel_;
```

Comment:
Probably scope creep or something for a follow-up... It would be great if we could use the first derivative of the position as velocity (we still have cmd_dis_old_).

Reply:
Was already on that. Will do it in next PR. I will also separate all commands for both interfaces to make joints in fake driver be controlled with different controllers. In current configuration it is only possible to move them all with position or velocity controller (not possible to move some of them with velocity and some of them with position controller)